### ISSUE TITLE
Hyper 1.0 support

### ISSUE THREAD

**Issue by davidpdrsn** [tokio-rs/axum#1325]
Hyper is aiming to ship 1.0 this year and in the process making a lot of changes to the public API which will impact axum. We can discuss the changes here.

Off the top of my head the biggest changes will probably be
- Hyper dropping `tower::Service` support since tower-service wont be 1.0 in time.
- Moving `hyper::Server` into hyper-util. I don't think we should have a public dependency on hyper-util since breaking releases would cascade into axum.
- Removing `hyper::Body` (or possibly moving it into hyper-util). I think we should consider giving axum-core its own body type. Relevant to https://github.com/tokio-rs/axum/issues/1110

I haven't thought much about this yet but once 0.6.0 is out we start that work.

Relevant links
- [Hyper 1.0 project board](https://github.com/orgs/hyperium/projects/1/views/6)
- [Hyper's roadmap](https://github.com/hyperium/hyper/blob/master/docs/ROADMAP.md)

**Comment by genusistimelord** [comment]
I guess the newer FromRequestParts and IntoResponseParts functions and State help Cover the loss of tower::Service. But this will heavily affect every single layer based library when and if you do update to the newest Hyper. So maybe we can improve and prepare all the Parts changes to make migration easier later on.  I am willing to switch my libraries over to using them so they can be used as Examples for other library makers to change theirs over too.

I'm guess hyper-util will be a faster change with more api breaking branch?

I think Axum should also have its own Body type. This would make it easier to Abstract the Body from different source and allow you more flexibility over them.

**Comment by davidpdrsn** [comment]
I've started the work in https://github.com/tokio-rs/axum/pull/1882 but it's currently blocked by tower-http getting updated. That is blocked by figuring out how to implement the (de?)compression middleware.

**Comment by davidpdrsn** [comment]
An update on where we're at with regards to hyper 1.0 support:

- Sean is planning to ship hyper 1.0 on November 15th 2023. This is super exciting! I can't guarantee that axum will be ready on the same day but that's my goal.
- The majority of the work to port axum is done. See https://github.com/tokio-rs/axum/pull/1882
- I've moved all issues, except this one, that were on the 0.7 milestone to a new 0.8 milestone. The focus for 0.7 is hyper 1.0 support. Other API changes will have to wait for the next major version.
- tower-http also needs to be updated. Most of that work is also done. See https://github.com/tower-rs/tower-http/pull/348. The only thing missing is compression and decompression middleware.

Finally most of the high level utilities that were previously in hyper (such as `hyper::Server`) are being moved to hyper-util. hyper-util is still early in development and thus pretty lean in terms of the conveniences it offers. That probably will hurt axum's usability at bit at first. I wanna encourage people to submit patches to hyper-util. I have started contributing myself (https://github.com/hyperium/hyper-util/pull/46).
